resolwe/flow/models/data.py

```diff
@@ -447,6 +476,56 @@ def delete(self, *args, **kwargs):
             if not storage.data.exists():
                 storage.delete()
 
+    def is_duplicate(self):
+        """Return True if data object is a duplicate."""
+        return bool(self.duplicated)
+
+    def _duplicate(self, contributor):
+        """Duplicate (make a copy)."""
+        if self.status not in [self.STATUS_DONE, self.STATUS_ERROR]:
+            raise ValidationError('Data object must have done or error status to be duplicated')
+
+        duplicate = Data.objects.get(id=self.id)
+        duplicate.pk = None
+        duplicate.slug = None
+        name_max_length = Data._meta.get_field('name').max_length
+        duplicate.name = 'Copy of {}'.format(self.name)[:name_max_length]
+        duplicate.contributor = contributor
+        duplicate.duplicated = datetime.datetime.now()
+        duplicate._perform_save(force_insert=True)  # pylint: disable=protected-access
+
+        # TODO
+        # Create duplication data migration history entry.
```

Comment:
This one, but please give directions how to do this.

Reply:
You just duplicate all [`DataMigrationHistory`](https://github.com/genialis/resolwe/blob/7ee372a9700a4f3fedd621cf551567354741d9aa/resolwe/flow/models/migrations.py#L37) entries.

Reply:
This was meant to log the relation between parent and child object. Before there was a new relation type `DataDependency`, but we discarded the idea. The question is how to log original -> duplicate.

Reply:
But since we're here, I'll also duplicate `DataMigrationHistory`.

Reply:
>But since we're here, I'll also duplicate DataMigrationHistory

This is already done [apparently](https://github.com/genialis/resolwe/pull/578/files#diff-aa73ad8a382e284ba1af2170c5e136caR524).

Reply:
Regarding the `DataMigrationHistory` entry.. what about something like

```py
DataMigrationHistory.objects.create(
    migration='identifier_please_suggest',
    data=duplicate,
    metadata={'parent_id': self.id},
)
```